[shuckycheese] iP

shuckycheese opened the pull request:
# James ChatBot
> "Computers are like Old Testament Gods; lots of rules and no mercy" - Joseph Campbell ([source](https://www.brainyquote.com/lists/topics/top-10-computers-quotes)) 

James frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

Here are some commands supported
1. list
2. bye
3. mark
4. unmark
5. todo
6. event
7. deadline
8. find

And it is __FREE!__ 😀 

Progress:
- [x] adding tasks
- [x] deletion and tracking of tasks
- [ ] filtering of tasks

Here is the entry point of the application:
``` java
public static void main(String[] args) {
        new James("data/james.txt").run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

shuckycheese committed:
renamed the chatbot and added a custom greeting and exit message

shuckycheese committed:
upgraded to accept commands and echo them back to the user

shuckycheese committed:
added a list feature that can list out all the recorded tasks

shuckycheese committed:
added a task class to be able to handle tasks better as well as a marking system for the tasks

shuckycheese committed:
added the ability to track 3 different subtypes of Task

shuckycheese committed:
added basic exception handling

shuckycheese committed:
added support for deleting tasks

shuckycheese committed:
Add testing

shuckycheese committed:
Add fucntion to load saved data from file

shuckycheese committed:
Add methods to properly update the data file whenever tasklist is updated

shuckycheese committed:
Merge branch 'branch-Level-7'

shuckycheese committed:
Update code to use LocalDateTime instead of strings

shuckycheese committed:
Fix parsing issue and display time

shuckycheese committed:
Merge branch 'branch-Level-8'

shuckycheese committed:
Refactor the code to make it more OOP

shuckycheese committed:
Merge branch 'branch-A-MoreOOP'

shuckycheese committed:
Fix field final declarations

shuckycheese committed:
Add classes to package James

shuckycheese committed:
Merge branch 'branch-A-Packages'

shuckycheese committed:
Merge branch 'add-gradle-support'

shuckycheese committed:
Change main application in build.gradle

shuckycheese committed:
Add JUnit testing for Storage and TaskList

shuckycheese committed:
Merge branch 'branch-A-JUnit'

shuckycheese committed:
Add JavaDoc comments for each class

shuckycheese committed:
Reformat code to comply with coding standards

shuckycheese committed:
Remove case indentation for switch block

shuckycheese committed:
Add find functionality

shuckycheese committed:
Add line breaks to lines that are too long

shuckycheese committed:
Merge branch 'branch-A-JavaDoc'

shuckycheese committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/James/CommandNotFoundException.java
#	src/main/java/James/Deadline.java
#	src/main/java/James/Event.java
#	src/main/java/James/James.java
#	src/main/java/James/JamesException.java
#	src/main/java/James/MissingDescriptionException.java
#	src/main/java/James/Parser.java
#	src/main/java/James/Storage.java
#	src/main/java/James/Task.java
#	src/main/java/James/TaskList.java
#	src/main/java/James/Todo.java
#	src/main/java/James/Ui.java

shuckycheese committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/James/Parser.java
#	src/main/java/James/Task.java

hm-yang1 commented on src/main/java/James/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+class Deadline extends Task {
+
+    private LocalDateTime deadline;
+    private final String PATTERN = "MMM d yyyy HHmm";
```
PATTERN used in multiple places. Could it somehow be abstracted?

hm-yang1 commented on src/main/java/James/Task.java:
```diff
@@ -0,0 +1,81 @@
+ */
+abstract class Task {
+    private String description;
+    private Boolean marked;
```
I would name it isMarked

hm-yang1 commented on src/main/java/James/CommandNotFoundException.java:
```diff
@@ -0,0 +1,17 @@
+ * Exception thrown when a command is not recognized.
+ * This exception is used to indicate that the command provided does not match any known commands.
+ */
+class CommandNotFoundException extends JamesException {
```
I like that there are multiple exceptions

hm-yang1 commented on src/test/java/james/StorageTest.java:
```diff
@@ -0,0 +1,53 @@
+        ArrayList<Task> taskList = new ArrayList<>();
+        Task todo = new Todo("Read a book", false);
+        Task deadline = new Deadline("Submit assignment", false, LocalDateTime.of(2024, 8, 31, 23, 59));
+        Task event = new Event("Project meeting", false, LocalDateTime.of(2024, 8, 30, 10, 0), LocalDateTime.of(2024, 8, 30, 12, 0));
```
Maybe add a line break here?

hm-yang1 commented on src/test/java/james/StorageTest.java:
```diff
@@ -0,0 +1,53 @@
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Is the best way to import?

hm-yang1 reviewed (commented):
Very good. Loved the detailed javadoc comments and custom exceptions.
Suggest minor cosmetic changes

shuckycheese committed:
Add support for GUI

shuckycheese committed:
Add new exception classes to enhance GUI

shuckycheese committed:
Merge branch 'branch-Level-10'

Nigeltzy commented on src/main/java/James/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+    /**
+     * Creates a new Deadline task.
+     *
+     * @param desc Description of the task
+     * @param mark Whether the task is marked as done
+     * @param deadline Deadline for the task
+     * @throws MissingDescriptionException If the description is missing
```
Perhaps these lines could end with a punctuation to remain consistent with other JavaDocs comments

Nigeltzy commented on src/main/java/James/Event.java:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * Creates a new Event task.
+     *
+     * @param desc Description of the task
+     * @param mark Whether the task is marked as done
+     * @param start Start time of the event
+     * @param end End time of the event
+     * @throws MissingDescriptionException If the description is missing
```
Similar punctuation comment raised previously

Nigeltzy commented on src/main/java/James/Parser.java:
```diff
@@ -0,0 +1,149 @@
+        case "delete":
+            if (words.length < 2) {
+                throw new IncorrectIdentifierException("Oops! Looks like you missed out the task number. " +
+                        "Please enter the task number after the command");
```
Similar operator line-break comment as mentioned previously

Nigeltzy commented on src/main/java/James/Task.java:
```diff
@@ -0,0 +1,81 @@
+     * Marks the task as done.
+     */
+    public void mark() {
+        this.marked = true;
```
While not mentioned in the [basic+intermediate] coding guidelines, perhaps you could consider applying the "use this keyword only when a field is shadowed by a method or constructor parameter" guideline under the Advanced Rule section.

Nigeltzy commented on src/main/java/James/Parser.java:
```diff
@@ -0,0 +1,149 @@
+        case "unmark":
+            if (words.length < 2) {
+                throw new IncorrectIdentifierException("Oops! Looks like you missed out the task number. " +
+                        "Please enter the task number after the command");
```
Perhaps the line break could be placed before the operator in accordance with the coding guidelines

Nigeltzy reviewed (commented):
LGTM. Only some minor cosmetic suggestions proposed.

shuckycheese committed:
Add assertions to code.

There are several methods that require instances of classes to be passed
in but in the off chance that they are null, the application will not work as
intended.

Add assertions for those methods to ensure class instances are not null.

Using assertions is preferable over exceptions as it is not a user error
if the instances passed in are null.

shuckycheese committed:
Improve code quality

Some of the methods in the code were too long (longer than 30 lines of
code) which affects the readability of the code.

Some boolean fields were also not named properly which may
cause confusion.

Broke down the long methods into smaller and shorter methods
such that it is easier to read.

Change the name of the boolean fields.

shuckycheese committed:
Merge pull request #1 from shuckycheese/branch-A-Assertions

Add assertions to code.

shuckycheese committed:
Merge branch 'master' into branch-A-CodeQuality

shuckycheese committed:
Merge pull request #2 from shuckycheese/branch-A-CodeQuality

Improve code quality

shuckycheese committed:
Improve find command

shuckycheese committed:
Merge branch 'branch-BCD-Extension'

shuckycheese committed:
Add line breaks for long lines

shuckycheese committed:
Add more exception handling

shuckycheese committed:
Improve list method

shuckycheese committed:
Add user guide

shuckycheese committed:
Add Ui picture to docs
